### ISSUE TITLE
Why is the convergence code zero even for models that don't give convergence warning?

### ISSUE THREAD

**Issue by JAQuent** [lme4/lme4#599]
Hey,
I apologise upfront if I am missing something here. Background: I want to run a simulation to see the distribution of p-values when I randomly shuffle my data. When I do this, I get a lot non-convergence warnings. Because of this, I want to directly access the convergence code of the model and save it with the simulation. After some digging in the S3 methods of the summary and print.summary, it seems to me that the convergence code is always zero even for models that seem to have converged. Here is my example:

```
set.seed(6) # Need to get convergence fail
cbpp$var <- rnorm(nrow(cbpp), 10, 10)

gm1 <- glmer(cbind(incidence, size - incidence) ~ period + (1 | herd),
            data = cbpp, family = binomial)

gm2 <- glmer(cbind(incidence, size - incidence) ~ period * var + (1 | herd),
             data = cbpp, family = binomial) # Fails to converge

gm1_sum <- summary(gm1)
gm2_sum <- summary(gm2)
```

```
> gm1_sum$optinfo$conv$opt
[1] 0
```

```
> gm2_sum$optinfo$conv$opt
[1] 0
```

I assumed that convergence code of zero is bad because of the message you get when it fails: 
```
> .prt.warn(gm2_sum$optinfo, summary = FALSE)
convergence code: 0
Model failed to converge with max|grad| = 0.00144205 (tol = 0.001, component 1)
```
Looking at this function though
```
function (optinfo, summary = FALSE, ...) 
{
    if (length(optinfo) == 0) 
        return()
    cc <- optinfo$conv$opt
    msgs <- unlist(optinfo$conv$lme4$messages)
    nmsgs <- length(msgs)
    warnings <- optinfo$warnings
    nwarnings <- length(warnings)
    if (cc > 0 || nmsgs > 0 || nwarnings > 0) {
        if (summary) {
            cat(sprintf("convergence code %d; %d optimizer warnings; %d lme4 warnings", 
                cc, nmsgs, nwarnings), "\n")
        }
        else {
            cat(sprintf("convergence code: %d", cc), msgs, 
                unlist(warnings), sep = "\n")
            cat("\n")
        }
    }
}
<bytecode: 0x00000236ac194350>
<environment: namespace:lme4>
```
It seems that based on convegence code only (`cc`), a message is only printed if it is over 0, which confuses me. In the documentation I couldn't find an explanation of the different convegence codes and their meaning.

**Comment by singmann** [comment]
I feel this is not really an `lme4` issue but a question about `lme4` that is better suited for `r-sig-mixed-models@r-project.org` (see https://github.com/lme4/lme4/#where-to-get-help).

Anyway, two things: 
First, I cannot reproduce the convergence warning with your code on current `R` and `lme4` on Windows. Maybe updating `lme4` removes the warning for you?

Second, the issue you see is that there are two different checks for convergence. The first check comes from the optimisation function itself. As `?glmerControl` reveals that is `optimizer = c("bobyqa", "Nelder_Mead")`. The optimiser (or optimisers in the `glmer` case) have some internal checks whether or not they believe the optimisation task has succeeded. If they think this has not happened they pass a convergence code other than 0 (which does not seem to happen here). 
Then in the next step, `lme4` runs some further checks on the solution returned by the optimiser. Whether or not these checks are run can also be controlled via `glmerControl`. In particular, it checks the gradient (`check.conv.grad`) and hessian (`check.conv.hess`) of the solution and whether it is singular (`check.conv.singular`). In the present case the first of those (the gradient check) seems to fail.
As these two checks (from the optimiser and for the gradient of the solution) are independent, you see the discrepancy.

**Comment by JAQuent** [comment]
Thank you for the explanation. This does make sense. I think it might be worth adding a feasture so that there is one consistent convergence code that you can access in the structure or least clarify that code of 0 is not the problem as it's return by the print.summary. While trying to find a solution to this problem I stumpled upon several [posts ](https://stackoverflow.com/questions/60002676/lme4-handling-lmer-convergence-code-0) that show pople think a convergence code of 0 is problematic and not something like 1, 10, 51, 52 from optim. This also confused me. 

For reproducibility of the problem. Here is my R session:
```
R version 3.6.2 (2019-12-12)
Platform: x86_64-w64-mingw32/x64 (64-bit)
Running under: Windows 10 x64 (build 17763)

Matrix products: default

Random number generation:
 RNG:     Mersenne-Twister 
 Normal:  Inversion 
 Sample:  Rounding 
 
locale:
[1] LC_COLLATE=English_United Kingdom.1252  LC_CTYPE=English_United Kingdom.1252    LC_MONETARY=English_United Kingdom.1252 LC_NUMERIC=C                           
[5] LC_TIME=English_United Kingdom.1252    

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] lme4_1.1-21   Matrix_1.2-18

loaded via a namespace (and not attached):
 [1] minqa_1.2.4     MASS_7.3-51.4   compiler_3.6.2  tools_3.6.2     Rcpp_1.0.3      splines_3.6.2   nlme_3.1-142    grid_3.6.2      nloptr_1.2.2.1  boot_1.3-23    
[11] lattice_0.20-38

I just installed lme4 to be sure it's the current one.
```

**Comment by singmann** [comment]
You might have tried to install the latest lme4 version, but that does not seem to have worked. The latest version is `1.1-23` and not your `1.1-21`. See: https://cran.r-project.org/package=lme4

**Comment by JAQuent** [comment]
You are right with version 1.1-23 seed 6 is not reproducible. I found this works:

```
set.seed(14)
cbpp$var <- rnorm(nrow(cbpp), 10, 10)

gm1 <- glmer(cbind(incidence, size - incidence) ~ period + (1 | herd),
             data = cbpp, family = binomial)

gm2 <- glmer(cbind(incidence, size - incidence) ~ period * var + (1 | herd),
             data = cbpp, family = binomial)

gm1_sum <- summary(gm1)
gm2_sum <- summary(gm2)
```

**Comment by bbolker** [comment]
I can think of some ways to enhance the printed message (by adding @optinfo$optimizer and possibly adding "(OK)" for conv code 0?  Thought I could use "msgs", which is a component of `optim` output explaining the convergence code, but I'm not sure that will be available most of the time ...

**Comment by JAQuent** [comment]
I think this would be helpful for instance something like 
```
optimiser (): convergence code: 0
lme4: Model failed to converge with max|grad| = 0.00144205 (tol = 0.001, component 1)
```

**Comment by bbolker** [comment]
How about

> optimizer (Nelder_Mead) convergence code: 0

It will be useful (but, it turns out, a little harder) to return the more detailed messages that the individual optimizers pass back about the meanings of their convergence codes ...

**Comment by JAQuent** [comment]
This would help already!

**Comment by bbolker** [comment]
resolved in 2dcda14
